Add out of space prevention mechanisms

paszkow opened the pull request:
When a scaling out is delayed or fails, it is crucial to ensure that clusters remain operational
and recoverable even under extreme conditions. To achieve this, the following proactive measures
are implemented:
- reject writes
      - includes: inserts, updates, deletes, counter updates, hints, read+repair and lwt writes
      - applicable to: user tables, views, CDC log, audit, cql tracing
- stop running compactions/repairs and prevent from starting new ones
- reject incoming tablet migrations

The aforementioned mechanisms are automatically enabled when node's disk utilization reaches
the critical level (default: 98%) and disabled when the utilization drop below the threshold.

Apart from that, the series add tests that require mounted volumes to simulate out of space.
The paths to the volumes can be provided using the a pytest argument, i.e.  `--space-limited-dirs`.
When not provided, tests are skipped.

Test scenarios:

1. Start a cluster and write data until one of the nodes reaches 90% of the disk utilization
2. Perform an **operation** that would take the nodes over 100%
3. The nodes should not exceed the critical disk utilization (98% by default)
4. Scale out the cluster by adding one node per rack
5. Retry or wait for the **operation** from step 2

The **operation** is: writing data, running compactions, building materialized views, running repair, 
migrating tablets (caused by RF change, decommission).

The test is successful, if no nodes run out of space, the **operation** from step 2 is
aborted/paused/timed out and the **operation** from step 5 is successful.

`perf-simple-query --smp 1 -m 1G` results obtained for fixed 400MHz frequency:

Read path (before)

```
instructions_per_op:
	mean=   39661.51 standard-deviation=34.53
	median= 39655.39 median-absolute-deviation=23.33
	maximum=39708.71 minimum=39622.61
```

Read path (after)

```
instructions_per_op:
	mean=   39691.68 standard-deviation=34.54
	median= 39683.14 median-absolute-deviation=11.94
	maximum=39749.32 minimum=39656.63
```

Write path (before):

```
instructions_per_op:
	mean=   50942.86 standard-deviation=97.69
	median= 50974.11 median-absolute-deviation=34.25
	maximum=51019.23 minimum=50771.60
```

Write path (after):

```
instructions_per_op:
	mean=   51000.15 standard-deviation=115.04
	median= 51043.93 median-absolute-deviation=52.19
	maximum=51065.81 minimum=50795.00
```

Fixes: https://github.com/scylladb/scylladb/issues/14067
Refs: https://github.com/scylladb/scylladb/issues/2871

No backport, as it is a new feature.

bhalevy commented on db/config.cc:
```diff
@@ -1456,6 +1456,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , disk_space_monitor_normal_polling_interval_in_seconds(this, "disk_space_monitor_normal_polling_interval_in_seconds", value_status::Used, 10, "Disk-space polling interval while below polling threshold")
     , disk_space_monitor_high_polling_interval_in_seconds(this, "disk_space_monitor_high_polling_interval_in_seconds", value_status::Used, 1, "Disk-space polling interval at or above polling threshold")
     , disk_space_monitor_polling_interval_threshold(this, "disk_space_monitor_polling_interval_threshold", value_status::Used, 0.9, "Disk-space polling threshold. Polling interval is increased when disk utilization is greater than or equal to this threshold")
+    , critical_disk_utilization_level(this, "critical_disk_utilization_level", liveness::LiveUpdate, value_status::Used, 0.98, "Disk utilization level above which all mutation writes to user table are rejected to prevent out-of-space errors")
```
Let's have a more general description and point to the documentation for more details about what the critical level controls.

bhalevy replied:
Let's consider either a .md document (in docs/dev) or a user visible .rst doc for user-oriented documentation.

Cc @annastuchlik

bhalevy commented on test/boost/database_test.cc:
```diff
@@ -56,6 +57,18 @@
 using namespace std::chrono_literals;
 using namespace sstables;
 
+class set_database_in_critical_disk_utilization_mode_guard {
```
Let's add comment that it requires a seastar thread.

bhalevy replied:
In second thought, I think we should change  the counter-based method.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1147,6 +1147,7 @@ future<> compaction_manager::drain() {
     }
     // Stop ongoing compactions, if the request has not been sent already and wait for them to stop.
     co_await stop_ongoing_compactions("drain");
+    _compaction_submission_timer.cancel();
```
can be submitted independently (and be backported)

bhalevy replied:
and make sure the dependency on that PR is mentioned in this one.

bhalevy replied:
https://github.com/scylladb/scylladb/pull/24505

bhalevy commented on replica/CMakeLists.txt:
```diff
@@ -1,24 +1,24 @@
-    wasmtime_bindings
+    db
     Seastar::seastar
+    wasmtime_bindings
```
Should be submitted independently

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1035,6 +1035,13 @@ void compaction_manager::register_metrics() {
 void compaction_manager::enable() {
     SCYLLA_ASSERT(_state == state::none || _state == state::disabled);
+
+    --_disabled_state_count;
```
Idea: change the `enabled` / `disabled` state to use an auxiliary counter
and leave only `none|running|stopped` in `compaction_manager::state` enum class.
And then disabled enabled state would be valid only in the `running` state and be determined by a counter.

paszkow replied:
Good idea. I will check if `stopped` is used only after a service is stopped, which should according to the documentation. Then we would get rid of all these states as the service cannot be restarted once stopped.

paszkow replied:
Done.

tgrabiec commented on service/topology_coordinator.cc:
```diff
@@ -1399,17 +1399,14 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-                                        .build());
-                            }
-                            break;
+                        if (do_barrier()) {
```
You could make a decision based on information in load_stats, which eventually propagates to the topology coordinator.

tgrabiec replied:
It's more reliable than trying to match error messages.

paszkow replied:
@bhalevy Tomek's comment was to tablet streaming. There we need to differentiate the source of the failure. 

Since you mentioned repair. I do not fail the entire repair operation but just a single tablet repair. The error is propagated to the `topology_coordinator` and the tablet is moved to the `end_repair` state. In master, we do not remove the repair_task_info content from system.tablets and thus tablet repair is retried until succeeds.

So my approach is to delete it from system.tablets. Then indeed the tablet repair is interrupted but I need to check the consequences of this approach.

tgrabiec replied:
> Definitely, but that requires to expand it. Capacity is not enough. We would need the current disk utilization per host. Then I agree using that would be better than parsing an error message.

We can expand it. If we want to make replica the decision-maker, we can send back only a flag which tells if the out-of-space rejection is on. Then there is no risk of issues related to differences in opinion.

bhalevy replied:
If we extend the rpc protocol, I'd prefer to return a well defined error code and possibly also error category to help distinguish between retry-able errors to others rather than a boolean out-of-space flag, so it could be used in the future for other error conditions as well.

cezarmoise commented:
In this case, it switched on and off 17 times in 2 minutes.
```
Jun 11 15:39:35.344884 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:39:36.344947 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 96.9%%
Jun 11 15:40:42.357338 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:40:43.358759 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.5%%
Jun 11 15:41:05.361540 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:06.362483 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.7%%
Jun 11 15:41:21.365683 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:22.367268 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.8%%
Jun 11 15:41:29.367955 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:30.368561 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.9%%
Jun 11 15:41:35.371201 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:36.370886 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.9%%
Jun 11 15:41:40.372116 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:41.372155 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:43.373375 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:44.373329 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
Jun 11 15:41:45.374295 oos-test-feature--db-node-be8fbad8-1 scylla[5461]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
```

Maybe a mechanism where it enables at 98% but disables at 97% would be a good idea.
I think I remember something like this being discussed.

paszkow commented:
@cezarmoise Good findings. Turns out the `nodetool cluster repair` uses a different code path. I have it fixed but I am afraid that we get to the similar problem as with streaming. I abort repair on any error not just due to critical disk utilization.

bhalevy commented on replica/database.cc:
```diff
@@ -735,6 +739,21 @@ do_parse_schema_tables(distributed<service::storage_proxy>& proxy, const sstring
+        } else if (db._critical_disk_utilization_mode_count > 0) {
+            --db._critical_disk_utilization_mode_count;
+            if (db._critical_disk_utilization_mode_count > 0) {
+                dblog.warn("Database is still in critical disk utilization mode, requires {} more call(s) to disable it",
```
Why is this a warning?  What should the user do with it?
It seems like it should be a debug-level message essentially.
If you're concerned about a bug we can periodically check here the disk_space_monitor.
Also, it may be beneficial to log a INFO message when entering and when exiting the critical disk utilization mode.

That said, why do we need to manage a state on the database rather than having a predicate that checks the disk_space_monitor against the configured threshold?
How would this state behave if the threshold is reconfigured and e.g. the database was in critical_disk_utilization_mode and after reconfig it isn't?

paszkow replied:
> Why is this a warning? What should the user do with it? It seems like it should be a debug-level message essentially. If you're concerned about a bug we can periodically check here the disk_space_monitor. Also, it may be beneficial to log a INFO message when entering and when exiting the critical disk utilization mode.

You are right. A user cannot do anything. Will change the level to debug.

> That said, why do we need to manage a state on the database rather than having a predicate that checks the disk_space_monitor against the configured threshold? 

The controller registers a callback to the `disk_space_monitor` and checks the current utilization against the configured threshold. Then it notifies relevant parties about reaching critical disk utilization. This way you keep the logic in the exactly one place, the controller. Otherwise, you will have code duplication because the same logic would have to be implemented in database, repair_service, compaction_manager.

The method was initially called `reject_user_table_writes()` that I later renamed. There I saw, similarly to `compaction_manager::drain()`, that if triggered by other callers than just the controller, we need to keep track on how many times something was enabled/disabled. In this case, we could remove the counter.

> How would this state behave if the threshold is reconfigured and e.g. the database was in critical_disk_utilization_mode and after reconfig it isn't?

If you reconfigure a threshold, the space monitor will get triggered. Then an appropriate callback from the controller will be called. Depending on the value, it will either drop the limitation or not. It works perfectly fine.

bhalevy commented on replica/database.cc:
```diff
@@ -899,6 +918,15 @@ sstables::sstables_manager& database::get_sstables_manager(const schema& s) cons
+    // audit - is not an internal keyspace in the sens of data_dictionary::keyspace::is_internal
+    // cql tracing - is part of system_traces that is an internal keyspace but we want to reject
+    return db.is_in_critical_disk_utilization_mode() &&
+        (!db.as_data_dictionary().find_keyspace(s.ks_name()).is_internal() || s.ks_name() == tracing::trace_keyspace_helper::KEYSPACE_NAME);
```
find_keyspace searches the keyspace name and this is done on the hot write path, for every write.
Since `find_column_family` is called anyhow later on, we can call it earlier and tag the tables in `database::tables_metadata::add_table_helper` based on a predicate for keyspaces (for `is_internal() || s.ks_name() == tracing::trace_keyspace_helper::KEYSPACE_NAME`) so we can have a `tbl.shall_reject_writes_in_critical_disk_utilization_mode()` predicate on the table itself.

bhalevy replied:
A back-pointer from the table to the keyspace object could be useful too, but I'm not sure how easy it would be to implement one.
Alternatively, we can also maintain something like `std::unordered_map<table_id, lw_shared_ptr<keyspace>> replica::database::_keyspace_by_table_id` and implement `std::tuple<table&, keyspace&> find_table_and_keyspace(table_id)` that will return both, but it will cost another unordered_map search,

paszkow replied:
I will have a look at it.

paszkow replied:
I tried to follow the second approach that I liked more, but the change became quite intrusive. Eventually, after fighting with issues, I turned towards the first suggestion and got implemented.

bhalevy commented on replica/database.hh:
```diff
@@ -1574,6 +1575,7 @@ private:
     compaction_manager& _compaction_manager;
     seastar::metrics::metric_groups _metrics;
     bool _enable_incremental_backups = false;
+    uint8_t _critical_disk_utilization_mode_count = 0;
```
uint8_t seems too narrow if the counter leaks due to a bug.
I suggest to change strategy and consult the disk_space_monitor and the configured threshold instead.
Or to "cache" the state and update it periodically via a time that will poll the above data + config value.
(Can be trigger on live update as well)

bhalevy commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,80 @@
+    {}
+
+future<> out_of_space_controller::stop() {
+    logger.info("Controller stopped");
```
Why do we need this message?

bhalevy replied:
Should we cancel the subscription and wait for its callback to be done in case its running now?
(Can hold a gate on entry)

paszkow replied:
Removed

bhalevy commented on replica/out_of_space_controller.hh:
```diff
@@ -0,0 +1,57 @@
+    struct config {
+        sharded<database>& db;
+        sharded<compaction_manager>& cm;
+        sharded<repair_service>& rs;
```
Those are usually passed as parameters to the constructor rather than in its configuration.

bhalevy commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,80 @@
+    auto cm_fh = disabled
+                ? [](compaction_manager& cm) { return cm.drain(); }
+                : [](compaction_manager& cm) { cm.enable(); return make_ready_future<>(); };
+    co_await cm.invoke_on_all(cm_fh);
```
Not sure we need cm_fh.
It'd be straight forward to do something like:
```c++
    if (disabled) {
        co_await cm.invoke_on_all([] (compaction_manager& cm) { return cm.drain(); });
    } else {
        co_await cm.invoke_on_all([] (compaction_manager& cm) { cm.enable(); });
    } 
```

bhalevy commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,80 @@
+    logger.info("{} repair service", disabled ? "disabling" : "enabling");
+    auto rs_fh = disabled
+                ? [](repair_service& rs) { return rs.drain(); }
+                : [](repair_service& rs) { rs.enable(); return make_ready_future<>(); };
```
ditto

bhalevy commented on main.cc:
```diff
@@ -2216,6 +2218,19 @@ sharded<locator::shared_token_metadata> token_metadata;
+            };
+            out_of_space_controller_shard0.emplace(std::move(out_of_space_controller_cfg), *disk_space_monitor_shard0, feature_service.local(),
+                                                       stop_signal.as_local_abort_source());
+            auto stop_streaming_controller = defer_verbose_shutdown("write throttling controller", [&] {
```
This produces a shut down log message as well, so I don't think the stop method itself needs to log anything by default.

bhalevy commented on test/pylib/scylla_cluster.py:
```diff
@@ -298,7 +298,17 @@ def __init__(self, mode: str, exe: str, vardir: str | pathlib.Path,
         shortname = f"scylla-{f'{xdist_worker_id}-' if xdist_worker_id else ''}{self.server_id}"
-        self.workdir = self.vardir / shortname
+
+        workdir = self.vardir / shortname
```
Please open issue and send as independent PR (and consider backport I guess to all live versions)

paszkow replied:
We do not need a backport. Existing tests do not need to overwrite the workdir except those that we introduce. Hence I put it as part of this PR and not a separate one.

bhalevy commented on test.py:
```diff
@@ -200,6 +200,10 @@ def parse_cmd_line() -> argparse.Namespace:
+    parser.add_argument("--pytest-arg", action='store', type=str,
+                        default=None, dest="pytest_arg",
+                        help="Additional command line arguments to pass to "
+                        "pytest, for example ./test.py --pytest-arg=\"-v -x\"")
```
Can be sent as an independent enhancement so to expedite it to master

paszkow replied:
Sent as a separate PR.

paszkow commented on idl/replica_exception.idl.hh:
```diff
@@ -23,12 +23,17 @@ class stale_topology_exception {
 };
 
+class critical_disk_utilization_exception {
+    sstring failed_action();
```
We do not. I just followed how other classes were written. I can remove it.

paszkow replied:
> Why do we need this method? It's not used in the patch that introduces it.

Turns out we do need it. It is not used directly by us, by it is used by the code generator to properly initialize the exception on the coordinator side when sent by replica.

paszkow commented:
In v2:

- Resolve minor styling issues
- Tag tables eligible to reject user table writes when added to database. This way we avoid additional search for a keyspace on every write just to check if the write can be rejected or not.
- Extended load_stats to contain information on critical disk utilization per node. This information is then used to check if tablet migrations shall be rejected. We want to reject migration only on critical disk utilization errors. On any other errors, it shall retry.
- Reject tablet rebuilds on the critical disk utilization. The tests showed and this path was not covered correctly. Now it's fixed.

bhalevy commented on replica/database.cc:
```diff
@@ -735,6 +745,22 @@ do_parse_schema_tables(distributed<service::storage_proxy>& proxy, const sstring
+    return sharded_db.invoke_on_all([enabled] (replica::database& db) {
+        dblog.info("Setting critical disk utilization mode: {}", enabled);
+        if (enabled) {
+            ++db._critical_disk_utilization_mode_count;
```
Need intrrnal error check against uint8_t wraparound

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1034,8 +1034,14 @@ void compaction_manager::register_metrics() {
-    _state = state::enabled;
+    SCYLLA_ASSERT(_state == state::none || _state == state::running);
+
+    if (_disabled_state_count > 0 && --_disabled_state_count > 0) {
```
Protect against underflow with on_internal_error_noexcept

paszkow replied:
Already protected with the first `_disabled_state_count > 0` check. The second check is to print a debug log.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1034,8 +1034,14 @@ void compaction_manager::register_metrics() {
+    SCYLLA_ASSERT(_state == state::none || _state == state::running);
+
+    if (_disabled_state_count > 0 && --_disabled_state_count > 0) {
+        cmlog.warn("Compaction manager is still disabled, requires {} more call(s) to enable()", _disabled_state_count);
```
nit: not sure that a warning is needed if this can happen  under normal circumstances

paszkow replied:
Should have been debug as for the remaining services.

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -99,16 +99,13 @@ private:
+    // running: running, started and enabled at least once. Whether new compactions are accepted or not is determined by the counter
+    enum class state { none, stopped, running };
     state _state = state::none;
+    uint8_t _disabled_state_count = 1;
```
nit: please add comment explaining why we start as disabled and where the compaction managrr is enabled.

That said, I think we can initialize the counter to 0 and decrease it in enable() only in the `running` state.

bhalevy commented on repair/repair.cc:
```diff
@@ -1146,6 +1146,10 @@ future<> repair::shard_repair_task_impl::run() {
 // itself does very little (mainly tell other nodes and CPUs what to do).
 future<int> repair_service::do_repair_start(gms::gossip_address_map& addr_map, sstring keyspace, std::unordered_map<sstring, sstring> options_map) {
     get_repair_module().check_in_shutdown();
+    if (_state != state::running || _disabled_state_count > 0) {
```
this logic repeats in several places, let's add a method like is_disabled()

bhalevy commented on repair/repair.cc:
```diff
@@ -1146,6 +1146,10 @@ future<> repair::shard_repair_task_impl::run() {
 future<int> repair_service::do_repair_start(gms::gossip_address_map& addr_map, sstring keyspace, std::unordered_map<sstring, sstring> options_map) {
     get_repair_module().check_in_shutdown();
+    if (_state != state::running || _disabled_state_count > 0) {
+        co_return co_await make_exception_future<int>(std::runtime_error("Repair service is disabled. No repairs will be started until it's re-enabled"));
```
use coroutine::return_exception

paszkow commented on gms/feature_service.hh:
```diff
@@ -171,6 +171,7 @@ public:
     gms::feature repair_based_tablet_rebuild { *this, "REPAIR_BASED_TABLET_REBUILD"sv };
     gms::feature enforced_raft_rpc_scheduling_group { *this, "ENFORCED_RAFT_RPC_SCHEDULING_GROUP"sv };
     gms::feature load_and_stream_abort_rpc_message { *this, "LOAD_AND_STREAM_ABORT_RPC_MESSAGE"sv };
+    gms::feature out_of_space_prevention { *this, "OUT_OF_SPACE_PREVENTION"sv };
```
It's been removed completely.

paszkow commented:
In v3:

- Rebased
- Removed the feature flag
- Replaced `uint8_t` with `uint32_t` for counters
- Lowered the severity of certain logs

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -629,7 +629,7 @@ future<compaction_manager::compaction_stats_opt> compaction_manager::perform_com
 
 std::optional<gate::holder> compaction_manager::start_compaction(table_state& t) {
-    if (_state != state::enabled) {
+    if (_state != state::running || _disabled_state_count > 0) {
```
This condition: `(_state != state::running || _disabled_state_count > 0)` is repeated in many places.
How about defining it once, in a predicate method?

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1034,8 +1034,14 @@ void compaction_manager::register_metrics() {
+        return;
+    }
+
+    _state = state::running;
```
I'd consider the following so that `_disabled_state_count` can be initialized to 0:
```c++
if (_state != state::running) {
    _state = state::running;
} else if (_disabled_state_count > 0 && --_disabled_state_count > 0) {
    cmlog.debug("Compaction manager is still disabled, requires {} more call(s) to enable()", _disabled_state_count);
    return;
}
```

bhalevy commented on repair/row_level.hh:
```diff
@@ -112,6 +112,21 @@ class repair_service : public seastar::peering_sharded_service<repair_service> {
+    state _state = state::none;
+    uint32_t _disabled_state_count = 0;
+
+    bool is_disabled() const { return _state != state::running || _disabled_state_count > 0; }
```
Same can be defined in compaction_manager

paszkow replied:
Yes, we can.

bhalevy commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,84 @@
+
+future<void> set_database_in_critical_disk_utilization_mode(sharded<database>& db, bool enabled)
+{
+    logger.info("{} database critical disk utilization mode (rejecting user writes)", enabled ? "enabling" : "disabling");
```
We now do more than rejecting user writes.
Do we want to list all actions taken, none, or some?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | i-0e974296bfb214db2 |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/)<br> :small_blue_diamond: boost/database_test<br> :small_blue_diamond: cluster/storage/test_out_of_space_prevention | :white_check_mark: | spider5.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/)<br> :small_blue_diamond: boost/database_test<br> :small_blue_diamond: cluster/storage/test_out_of_space_prevention | :white_check_mark: | spider5.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/) | :white_check_mark: | spider3.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17902/)<br> :small_blue_diamond: boost/database_test<br> :small_blue_diamond: cluster/storage/test_out_of_space_prevention | :white_check_mark: | spider3.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5701/) | :white_check_mark: |  |

#### Build Details:
- Duration: 2 hr 36 min

paszkow commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,84 @@
+
+        co_await set_database_in_critical_disk_utilization_mode(_db, _critical_disk_utilization_threshold_reached);
+        co_await set_compaction_manager_disabled(_cm, _critical_disk_utilization_threshold_reached);
+        co_await set_repair_service_disabled(_rs, _critical_disk_utilization_threshold_reached);
```
Each lower level service has `drain` and `enable` methods or their counterparts. My idea was to have a controller that calls them when the state changes. This way these services stay the controller unaware. The entire logic stays in one place.

We could do the opposite and register callbacks but how is that better? Neither of the services shall know anything about the out_of_space_controller. They should not care if the disk is utilized in 10% or 100%.

paszkow replied:
> Also, do we need some hysteresis to prevent jitter around the threshold?

We might need it. In one of the recent tests Cezar run, we reached and dropped below the critical disk utilization level multiple times in one minute. On the other hand, is that frequent enough to handle it? If we want to have a factor (e.g. 0.5%), it cannot be fixed. For bigger instances it could be quite a lot of space.
```
3825:2025-06-25T16:25:17.724+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
18337:2025-06-25T16:25:19.565+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 96.7%%
18406:2025-06-25T16:26:39.741+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
33013:2025-06-25T16:26:41.462+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.4%%
33067:2025-06-25T16:27:09.751+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
47185:2025-06-25T16:27:11.431+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.7%%
47202:2025-06-25T16:27:24.754+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
61434:2025-06-25T16:27:26.410+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.8%%
61473:2025-06-25T16:27:34.757+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
70825:2025-06-25T16:27:35.884+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.9%%
70835:2025-06-25T16:27:40.756+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
85403:2025-06-25T16:27:42.493+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 97.9%%
85456:2025-06-25T16:27:45.600+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
99395:2025-06-25T16:27:47.338+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
99411:2025-06-25T16:27:48.507+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
108581:2025-06-25T16:27:49.873+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Dropped below the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
108590:2025-06-25T16:27:50.759+00:00 oos-test-feature--db-node-2cce87c6-3     !INFO | scylla[5258]:  [shard 0:main] out_of_space_controller - Reached the critical disk utilization level (98.0%%). Current disk utilization 98.0%%
```

bhalevy replied:
> > Also, do we need some hysteresis to prevent jitter around the threshold?
> 
> We might need it. In one of the recent tests Cezar run, we reached and dropped below the critical disk utilization level multiple times in one minute. On the other hand, is that frequent enough to handle it? If we want to have a factor (e.g. 0.5%), it cannot be fixed. For bigger instances it could be quite a lot of space.
> 

It is reasonable to stay in the critical state for 10 seconds minimum and then to reevaluate the state.  Let's kust keep the hysteresis factor in mind for now.

bhalevy commented on replica/database.cc:
```diff
@@ -1995,6 +2026,11 @@ future<> database::do_apply(schema_ptr s, const frozen_mutation& m, tracing::tra
     auto& cf = find_column_family(uuid);
 
+    if (is_in_critical_disk_utilization_mode() && cf.is_eligible_to_write_rejection_on_critical_disk_utilization()) {
+        ++_stats->total_writes_rejected_due_to_out_of_space_prevention;
```
Why not call `update_write_metrics_for_rejected_writes` here too?

paszkow replied:
The function you mentioned updates the number of write attempts, write failures + failures due to out-of-space problems. This `if`-statement has been moved to `do_apply` and in this function we already update the remaining two metrics:
```
++_stats->total_writes;
// assume failure until proven otherwise
auto update_writes_failed = defer([&] { ++_stats->total_writes_failed; });
```

bhalevy commented on locator/tablets.hh:
```diff
@@ -399,6 +399,9 @@ struct load_stats {
     std::unordered_map<host_id, uint64_t> capacity;
 
+    // Critical disk utilization check for each host.
+    std::optional<std::unordered_map<locator::host_id, bool>> critical_disk_utilization;
```
Why does it need to be optional?
Is it any different from an empty map when initialized?

bhalevy commented on locator/tablets.cc:
```diff
@@ -728,6 +728,16 @@ load_stats& load_stats::operator+=(const load_stats& s) {
     }
+    if (s.critical_disk_utilization) {
+        if (!critical_disk_utilization) {
+            critical_disk_utilization = s.critical_disk_utilization;
```
If the member wasn't optional than we'd need only the `else` loop below, simplifying this block.

paszkow replied:
Agree.

bhalevy commented on locator/tablets.cc:
```diff
@@ -728,6 +728,16 @@ load_stats& load_stats::operator+=(const load_stats& s) {
+            critical_disk_utilization = s.critical_disk_utilization;
+        } else {
+            for (auto& [host, cdu] : s.critical_disk_utilization.value()) {
+                critical_disk_utilization.value()[host] = cdu;
```
How about decommissioned/removed/replaced nodes?
Don't we need to delete their data?
Is it reset automatically on other events?

paszkow replied:
That is a good question. As I see, the `loads_stats` are created from scratch in `service::storage_service::load_stats_for_tablet_based_tables`. So eventually, the data for these nodes will not be present.

Even if we do have this outdated data for a while, we do not migrate tablets to the nodes you mentioned, don't we? @tgrabiec, please correct me if I am wrong here.

bhalevy commented on service/topology_coordinator.cc:
```diff
@@ -1423,8 +1423,14 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                                .or_else([&host=trinfo.pending_replica->host] {
+                                    return std::optional(decltype(locator::load_stats::critical_disk_utilization)::value_type({{host, false}}));
+                                })
+                                .value()[trinfo.pending_replica->host];
```
Here too I think that having critical_disk_utilization non-optional wold simplify the code, e.g.
```c++
const auto& util = _tablet_allocator.get_load_stats()->critical_disk_utilization;
auto it = util.find(trinfo.pending_replica->host);
bool critical_disk_utilization = it != util.end() && it->second;
```

bhalevy commented on streaming/stream_blob.cc:
```diff
@@ -179,6 +180,10 @@ future<> stream_blob_handler(replica::database& db,
             });
 
+            if (db.is_in_critical_disk_utilization_mode()) {
+                throw replica::critical_disk_utilization_exception("rejected streamed file");
```
please be more elaborate in exception string, e.g. "rejected streamed file due to critical disk utilization"

paszkow commented:
In v4:

- Rebase
- Remove unnecessary logging. Add logs, howmever, to drain/enable methods to compaction_manager and repair_service
- Adjust unit tests affected by the above change
- Remove optional from load_stats
- Fix coding style
- Start compaction_manager in enabled state instead of disabled

bhalevy commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,74 @@
+        co_await notify();
+    }))
+    {
+        dsm.trigger_poll();
```
Why is this needed if there are no subscribers yet?

paszkow replied:
I initially needed it for the unit tests but you are right. Most likely it's a left over after the re-write of the controller. I will double check and remove it.

bhalevy commented on replica/out_of_space_controller.hh:
```diff
@@ -0,0 +1,95 @@
+    };
+
+    using reached_critical_disk_utilization = bool_class<struct reached_critical_disk_utilization_tag>;
+    using subscription_callback_type = noncopyable_function<future<void> (reached_critical_disk_utilization)>;
```
nit: we typically use `future<>` rather than `future<void>`

bhalevy commented on replica/out_of_space_controller.hh:
```diff
@@ -0,0 +1,95 @@
+
+namespace replica {
+
+// Simple controller to notify database it is in the critical disk utilization zone. The action
```
nit: it's not only the database that is notified, this could be generalize to something like `notify subscribers, such as the database`

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -962,7 +962,14 @@ compaction_manager::compaction_manager(config cfg, abort_source& as, tasks::task
+        if (bool(critical_level_reached)) {
+            return container().invoke_on_all([] (compaction_manager& cm) { return cm.drain(); });
+        }
+        return container().invoke_on_all([] (compaction_manager& cm) { cm.enable(); });
```
Shouldn't `enable()` or `drain()` be called only if `state == running`?
Otherwise, if the callback is called early before the compaction manager was enabled the first time there is no need to drain() it, and enable() would be premature as well.

bhalevy commented on replica/out_of_space_controller.cc:
```diff
@@ -0,0 +1,74 @@
+
+auto out_of_space_controller::subscribe(subscription_callback_type cb) -> subscription {
+    auto sub = subscription(*this, std::move(cb));
+    _dsm.trigger_poll();
```
ok to trigger_poll here, my bad.  the poller runs in a background fiber that's being signaled so the subscription will be called back later, if needed, and after the object (synchronous) constructor returned successfully.

mykaul commented:
@paszkow - please resolve conflicts.

paszkow commented:
In v8:

- Rebased and conflict resolution
- Address all the recent review comments
- Reject incoming row diffs in the row-level repair ( `repair_meta::put_row_diff_handler` ) to handle a situation where a tablet repair is initiated on a node with disk utilization below the critical threshold. This prevents the nodes with critical disk utilization to hit out of space errors.

mykaul commented:
Thanks @cezarmoise - good list. Anything missing?

paszkow commented:
Only rebase to fix conflicts.

avikivity commented:
Please incorporate the test description and test results into the cover letter. Think of it as a guide for the next one to touch this area - how would they know what to test?

avikivity commented on docs/troubleshooting/error-messages/critical-disk-utilization.rst:
```diff
@@ -0,0 +1,31 @@
+Solution
+^^^^^^^^
+
+Scale out a cluster. Adding new nodes, will bring the disk utilization down below the critical threshold and the prevention mechanisms will be deactivated.
```
s/a cluster/the cluster/.

Adding new nodes will help only if they are added in the same rack as the node that reached critical condition.

avikivity commented on replica/database.cc:
```diff
@@ -742,6 +756,22 @@ do_parse_schema_tables(distributed<service::storage_proxy>& proxy, const sstring
+        dblog.info("Setting critical disk utilization mode: {}", enabled);
+        if (enabled) {
+            ++db._critical_disk_utilization_mode_count;
+        } else if (db._critical_disk_utilization_mode_count > 0) {
```
This `if` is fishy, how can the increments and decrements be unmatched?

avikivity replied:
Then please drop the special handling, working around an impossible bug reduces confidence in the code.

avikivity commented on test/cluster/storage/README:
```diff
@@ -0,0 +1,24 @@
+--space-limited-dirs=testlog/dev/mnt/scylla-1;testlog/dev/mnt/scylla-2;testlog/dev/mnt/scylla-3
+
+or via test.py:
+--pytest-arg="--space-limited-dirs testlog/dev/mnt/scylla-1;testlog/dev/mnt/scylla-2;testlog/dev/mnt/scylla-3"
```
We can integrate them into the tests. Under dbuild, you're root and can do anything. Even outside dbuild, the following works:

```console
$ unshare -mr
# mount -t tmpfs -o size=100M foo /mnt
# df -h /mnt
Filesystem      Size  Used Avail Use% Mounted 
foo             100M     0  100M   0% /mnt
```

avikivity replied:
This is a different one, test.py creates the volumes itself. It works whether test.py is runs under dbuild or not.

avikivity replied:
And "appropriate volumes created" means no local reproducer for someone who isn't an expert at exactly what's going on here.

paszkow replied:
Yes, indeed. The content of the mount is only visible inside the namespace and the mounts themselves. Files and directories created outside the mounts, are visible everywhere. 

So we have access to a scylla.log file (created outside the mount) but the node's content is gone. Need to figure it out as the current behaviour of test.py is to keep data on tests failure or when `--save-log-on-success` is provided.

The auto mounting volumes works. Run BYO on jenkins that executed these tests and they pass.

avikivity commented:
Since we touch the data paths here, please add the results of `scylla perf-simple-query --smp 1 -m 1G` and `scylla perf-simple-query --smp 1 -m 1G --write`, before and after, in release mode. Only the intructions_per_op metric is stable enough to be interesting.

gleb-cloudius commented on docs/troubleshooting/error-messages/critical-disk-utilization.rst:
```diff
@@ -0,0 +1,31 @@
+^^^^^
+
+INSERT/UPDATE/DELETE cqlsh operations fail with ``Critical disk utilization: rejected write mutation`` error message.
+
```
Yes, LWT reads write into paxos table (though those writes ate tiny) and can even result in a repair in which case it will write into base table too.

gleb-cloudius replied:
But then regular select can also cause read repair which will write as well.

paszkow commented on idl/replica_exception.idl.hh:
```diff
@@ -23,12 +23,17 @@ class stale_topology_exception {
 class abort_requested_exception {
 };
```
Agh...I was probably copy-pasting incorrect thing...will fix the commit message.

paszkow replied:
Fixed the commit message.

paszkow commented on replica/database.hh:
```diff
@@ -1702,6 +1722,9 @@ public:
     void set_enable_incremental_backups(bool val) { _enable_incremental_backups = val; }
 
+    static future<> set_in_critical_disk_utilization_mode(sharded<database>& sharded_db, bool enabled);
+    bool is_in_critical_disk_utilization_mode() const { return _critical_disk_utilization_mode_count; }
```
Done

paszkow commented:
In v9:

- Clarified the documentation in `critical-disk-utilization.rst` file
- Fixed the typo in the commit message
- Added `[[unlikely]]` to `database::is_in_critical_disk_utilization_mode`
- Extended the cover letter with the test scenarios and the results from simple-perf-query

bhalevy commented:
@scylladb/scylla-maint please consider merging

mykaul commented:
@avikivity - please (re)review.

avikivity commented on test/lib/cql_test_env.cc:
```diff
@@ -161,6 +162,7 @@ class single_node_cql_env : public cql_test_env {
     sharded<locator::effective_replication_map_factory> _erm_factory;
     sharded<sstables::directory_semaphore> _sst_dir_semaphore;
     std::optional<utils::disk_space_monitor> _disk_space_monitor_shard0;
+    std::optional<replica::out_of_space_controller> _out_of_space_controller_shard0;
```
Maybe we should merge disk_space_monitor and out_of_space_controller. @bhalevy ?

bhalevy replied:
sounds like a good idea

avikivity commented:
Looks good apart from the requirement for externally supplied volumes. This doesn't work if different release branches need different volumes or if someone wants a local reproducer but doesn't exactly match what the README said.

test.py should run all the tests, whether from Jenkins or locally, and from dbuild or the host environment (provided all the dependencies are there).

The solution I outlined (with unshare -mr; mount) should work.

avikivity commented on test/cluster/storage/conftest.py:
```diff
@@ -0,0 +1,69 @@
+
+    if preserve_data:
+        for id, path in enumerate(volumes):
+            shutil.copytree(path, base.parent.parent / f"scylla-{hash}-{id}")
```
I hope that none of these tests fail because I don't want archives of filler sstables clogging up the reports.

paszkow replied:
It follows the exact same behavior of `test.py` for the remaining tests. 

But this is actually a good point. In my tests, in order to reach the critical disk utilization, I create random blob files of a certain size that might not be deleted in case of a test failure. I sent a new version that fixes that. Now I use 
```
with random_content_file(...):
    <code>
```
so the file is always removed no matter of the test result. Now when a test fails, we copy around a few MB of data only.

bhalevy commented on main.cc:
```diff
@@ -1173,6 +1174,15 @@ sharded<locator::shared_token_metadata> token_metadata;
             });
+            auto out_of_space_subscription = disk_space_monitor_shard0->subscribe(cfg->critical_disk_utilization_level, [&threhsold = cfg->critical_disk_utilization_level, &dsm = *disk_space_monitor_shard0] (auto threshold_reached) {
+                static constexpr auto msg_template = "{} the critical disk utilization level ({:.1f}%). Current disk utilization {:.1f}%";
+                if (bool(threshold_reached)) {
```
nit: is the explicit cast to bool needed?

paszkow replied:
Yes, because of the `explicit operator bool()`  in the `bool_cass`

avikivity replied:
It should not be needed, see Contextual Conversions in https://en.cppreference.com/w/cpp/language/implicit_conversion.html.

paszkow replied:
Correct. Fixed.

bhalevy commented on main.cc:
```diff
@@ -1173,6 +1174,15 @@ sharded<locator::shared_token_metadata> token_metadata;
             auto stop_dsm = defer_verbose_shutdown("disk space monitor", [&disk_space_monitor_shard0] {
                 disk_space_monitor_shard0->stop().get();
             });
+            auto out_of_space_subscription = disk_space_monitor_shard0->subscribe(cfg->critical_disk_utilization_level, [&threhsold = cfg->critical_disk_utilization_level, &dsm = *disk_space_monitor_shard0] (auto threshold_reached) {
```
nit: why `auto`? it wpuld be clearer to specify the real type here

paszkow replied:
Why do you need a real type here? The variable is unused. It is there only to ensure the subscription is not automatically removed. So `auto` is actually more clear to me in this case.

bhalevy replied:
> Why do you need a real type here? The variable is unused. It is there only to ensure the subscription is not automatically removed. So `auto` is actually more clear to me in this case.

I meant in `auto threshold_reached`, but it's not critical.

bhalevy commented on utils/disk_space_monitor.cc:
```diff
@@ -30,6 +30,28 @@ disk_space_monitor::disk_space_monitor(abort_source& as, std::filesystem::path d
+            if (constant_update || crossed_threshold) {
+                if ((threshold_reached == above_threshold::yes && sub._trigger_options.when_above_threshold) ||
+                    (threshold_reached == above_threshold::no && sub._trigger_options.when_below_threshold)) {
+                    co_await sub(std::move(threshold_reached));
```
what about yhe contant update case?
did I miss anything?
is there a test case for it?

paszkow commented:
v12:

- Dropped implicit bool cast in `if (bool(threshold_reached))`
- Corrected the flaky test
- Limited commitlog segment size and schema commitlog segment size as default 64MB and 128MB respectively are way too much for 100MB volumes used in tests.

paszkow commented:
> #### Failed Tests (6/82825):
> * [test_insert_failure_doesnt_report_success.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19026/testReport/junit/cluster.dtest.audit_test/TestCQLAudit/dev___test_py___dev___test_insert_failure_doesnt_report_success_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_insert_failure_doesnt_report_success.dev.1)
> * [test_insert_failure_doesnt_report_success.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19026/testReport/junit/cluster.dtest.audit_test/TestCQLAudit/dev___test_py___dev___test_insert_failure_doesnt_report_success_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_insert_failure_doesnt_report_success.dev.2)
> * [test_insert_failure_doesnt_report_success.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19026/testReport/junit/cluster.dtest.audit_test/TestCQLAudit/dev___test_py___dev___test_insert_failure_doesnt_report_success_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_insert_failure_doesnt_report_success.dev.3)

https://github.com/scylladb/scylladb/issues/25541

paszkow commented on test/cluster/storage/test_out_of_space_prevention.py:
```diff
@@ -0,0 +1,376 @@
+
+                    logger.info("Restart the node")
+                    await manager.server_restart(servers[0].server_id)
+                    await log.wait_for("database - Setting critical disk utilization mode: true", from_mark=mark, timeout=10)
```
Fixed

bhalevy commented on test/cluster/storage/suite.yaml:
If all the new unit tests pass, we can fix this as a follow-up after promotion.
If the series is dequeued for some reason, let's fix it before requeuing

xtrey replied:
Moreover, this way of executing test is breaking the flow with running test with pure `pytest` without `test.py`

xtrey replied:
https://github.com/scylladb/scylladb/tree/master/test/cluster/object_store I don't see any `suite.yaml` files in the object store directory. Definition what is suite and what is just directory is done by the file `suite.yaml`
Object store is a directory inside the cluster directory, where cluster is a suite and object_store just part of the suite cluster.

xtrey replied:
As a quick workaround, yes. But `test.py` and, that is more important, `pytest` do not have a support for the execution with `unshare`. 
BTW, your test should work without `unshare` in the toolchain, since we have root permission in it.

paszkow replied:
I have moved the tests outside the cluster suite as we agreed. Kept `unshare` for now and as a follow-up will get rid of it. Also as agreed.

yaronkaikov commented:
@avikivity This is breaking Gating, and i assume that the reason you queued it is due to a bug we have in CI (which we already fixed) that it was showing the wrong status :-( - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19075/ the last CI had multiple failures. 

Please dequeue

paszkow commented on test/boost/disk_space_monitor_test.cc:
```diff
@@ -0,0 +1,161 @@
+            poll_barrier.advance_and_await().get();
+        }
+
+        dsm.stop().get();
```
Done in the new version.

paszkow commented:
- Fixed the tests `test_node_restart_while_tablet_split`, `test_tablet_repair`. They were looking for logs on the incorrect nodes. The fix, asks which node is the topology coordinator and waits for the logs on this node
- The last failure reported `seastar::no_sharded_instance_exception` Looks like in the `debug` mode, the subscription was triggered before all local instances of a service were created. Moved subscribing to the disk monitor away from the constructors into `start` methods.

paszkow commented:
@scylladb/scylla-maint CI passed and this time it's not a lie. The tests truly passed. Please consider merging.

fruch commented:
this one just surfaced up in on of SCT weekly tests within nemesis `disrupt_nodetool_enospc`

it's fails the main load stress command with such a failure:
```
Validation failed:unable to load check data from the test store: Critical disk utilization: rejected write mutation (potentially executed: false)
```

we'll need some guidance of what do with those need errors (i.e. retries or something else)

avikivity commented:
> this one just surfaced up in on of SCT weekly tests within nemesis `disrupt_nodetool_enospc`
> 
> it's fails the main load stress command with such a failure:
> 
> ```
> Validation failed:unable to load check data from the test store: Critical disk utilization: rejected write mutation (potentially executed: false)
> ```
> 
> we'll need some guidance of what do with those need errors (i.e. retries or something else)

The change made the error recoverable.

Previously, one needed user intervention to recover, now it's automatic.

You can enhance the nemesis:
1. detect the new error message
2. remove the file that caused the out-of-space condition, and verify that the cluster recovers fully

We also need to verify that migration out of the affected node still works, but I believe we have unit tests for that.

fruch commented:
> > this one just surfaced up in on of SCT weekly tests within nemesis `disrupt_nodetool_enospc`
> > it's fails the main load stress command with such a failure:
> > ```
> > Validation failed:unable to load check data from the test store: Critical disk utilization: rejected write mutation (potentially executed: false)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we'll need some guidance of what do with those need errors (i.e. retries or something else)
> 
> The change made the error recoverable.
> 
> Previously, one needed user intervention to recover, now it's automatic.
> 
> You can enhance the nemesis:
> 
> 1. detect the new error message
> 2. remove the file that caused the out-of-space condition, and verify that the cluster recovers fully
> 
> We also need to verify that migration out of the affected node still works, but I believe we have unit tests for that.

nemesis is doing all what mentioned,

but it's new that we are getting this type of errors during CQL commands from the drivers, and we'll need to define the behavior should be in the stress tool for those error (for example gemini report it as failure, and not sure if it's retrying it, don't know what would every tool gonna do with this new error), I think it should be documented very clear for users as their code might not handle it as other known failure

bhalevy commented:
> If this feature required any client change behaviour it wasn't communicated and wasn't implemented.

It's a shame that CQL in general is under documented in many areas.
Do we have existing dicumentation about CQL error codes and how the client driver and application should treat them?

Cc @annastuchlik 

> In addition, enospc testing were disabled in SCT due to this feature cannot handling actual enospc.

We need tests that push nodes towards out-of-space and test this feature in action.
The expectation is that it should allow the cluster to scale out and writes and repair to suceed once nodes go out of the critical utilization danger zone.

paszkow commented:
> First, I don't understand why would the read repair errors will cause QUORUM breach when only one node is having space issues.

Because for RF=3, CL=QUORUM, the coordinator selects exactly 2 nodes where data/digest requests are sent to. If one of these two nodes does not have data that was requested (or have but old one), we have a digest mismatch. So the coordinator computes the differences and sends the update (a write query) to the nodes that participated in the read. If one of these two nodes is running out of space, it will reject the write. This write is executed with the same CL as the original read. The quorum of 2 is 2. That's why we see errors.

Normal writes are successful. We still have a quorum of nodes that acknowledged the write.  

> Second, it's a regression from the previous behavior.

I would not call it a regression. It's a desired change. We want to keep a node that is running out of space UP. So we can scale out, and the cluster can recover itself automatically. Migrate tablets away so the disk utilization drops.

annastuchlik commented:
> It's a shame that CQL in general is under documented in many areas. Do we have existing dicumentation about CQL error codes and how the client driver and application should treat them?

We don't have a full list. I'll open an issue to extend the docs with CQL error codes, but I don't think there's one source of truth I could use to address it myself. The dev team would have to work on that, too.

bhalevy commented:
> > > Second, it's a regression from the previous behavior.
> > 
> > 
> > I would not call it a regression. It's a desired change. We want to keep a node that is running out of space UP. So we can scale out, and the cluster can recover itself automatically. Migrate tablets away so the disk utilization drops.
> 
> With availability as a trade off, doesn't make sense to me...

This was beaten to death in the past so I don't want to revive the discussion all over again.  The gist of it is that if just a single node runs out of space and isolates itself, we can tolerate it and resort to manual processes to bring it back up and repair it.  But when running at 90% utilization as required to improve TCO, with the premise of tablets enabling agile scale out, there is a very real risk of multiple nodes running out of space if scale out is not fast enough.  So you must have some back pressure mechanism to prevent running OOS on multiple nodes, where loss of availability is a much more serious risk.

The mitigation plan we need but is not being worked on yet, unfortunately, is dynamic and predictive scale out in siren, that will extrapolate the growth rate and will prevent nodes reaching critical utilization by triggering scale out proactively, based on growth prediction rather than basing it on a static threshold that would work in most cases but not in the 1% freakish case of extreme data ingestion rate at 90% utilization, say during the beginning of backup where much of the capacity is locked to the snapshot and we race also with backup evicting sstables to object storage (we decided not to abort backup when reaching critical utilization).

roydahan commented:
> We believe that the test failure was caused by read-repair induced write which unfortunately has stricter consistency requirement currently (See #28125). We agree that the client should not experience an error in this case.
> But that said, it is an existing problem that got exacerbated by OOS prevention.

I understand all that.
Hence it's a regression.
And 2025.4 is exposed to that regression as long as this new feature is enabled.

avikivity commented:
> So coordinator of reads should avoid picking the replica that is in critical zone? Since such replica is rejecting writes and read-repair will be potentially triggered when this replica is picked for reads. Eventually we need to move OOS prevention to coordinator right? Migrations could be helpful to avoid it. In addition to signalling to an observer that scale out is needed.

OOS prevention cannot be done on the coordinator, it makes the decision too early.

OOS happens because provisioning/migration fail, we can't postulate that they will succeed. We start with the assumption that provisioning failed or migration failed and end up with an OOS node. OOS prevention has to make sure the node recovers quickly once provisioning/migration are fixed.
